File: C++Verifier/src/expressions/EqExpr.cpp
#ifndef EQEXPR_H
#define EQEXPR_H

#include <vector>
#include "../utils/utils.cpp"
#include "../DSFA/DSFA.h"

using namespace std;

class EqExpr: public BoolExpr {
    public:
        using BoolExpr::BoolExpr;

        void* evaluate(shared_ptr<Packet> p) {
            // cout << "In EqExpr evaluateA: " << toString() << endl;
            string leftValue = evalSingle(left, leftType, p);
            string rightValue = evalSingle(right, rightType, p);
            // cout << "Comparing: " << leftValue << " with: " << rightValue << endl;
            bool* ret = new bool;
            *ret = (leftValue == rightValue) ^ negated;
            // cout << "returned value: " << *ret << endl;
            return static_cast<void*>(ret);
        }

        bool evaluate(shared_ptr<Packet> p,  vector<string> constraints, vector<string> locationList, vector<string> variableList, int currentState, vector<shared_ptr<ConstraintTreeNode>> rootConstraints) {
            // cout << "In EqExpr evaluateB: " << toString() << endl;
            string leftValue = evalSingle(left, leftType, p, constraints, locationList, variableList);
            string rightValue = evalSingle(right, rightType, p, constraints, locationList, variableList);
            // cout << "Comparing: " << leftValue << " with: " << rightValue << endl;


            // Now parse variables.  Need to do this after the above so that we have the concrete int[] values
            if (leftType == VARIABLE || leftType == LOCATION_VAR) {
                int level;
                if (leftType == VARIABLE) {
                    level = locationList.size() + getIndexInVector(variableList, leftValue);
                } else {
                    level = getIndexInVector(locationList, leftValue);
                }

                if (level >= constraints.size() || constraints[level] == "") {
                    assert(rightValue != "");
                    DSFA::splitIfNew(constraints, 0, rightValue, currentState, rootConstraints);

                    return negated;
                } else {
                    leftValue = constraints[level];
                }

            } else if (rightType == VARIABLE || rightType == LOCATION_VAR) {
                int level;
                if (rightType == VARIABLE) {
                    level = locationList.size() + getIndexInVector(variableList, rightValue);
                } else {
                    level = getIndexInVector(locationList, rightValue);
                }

                if (level >= constraints.size() || constraints[level] == "") {
                    assert(leftValue != "");
                    DSFA::splitIfNew(constraints, 0, leftValue, currentState, rootConstraints);
                    return negated;
                } else {
                    rightValue = constraints[level];
                }
            } else if (leftValue == "" || rightValue == "") {
                throw std::runtime_error("Unknown Eq Expression fields");
            }

            return (leftValue == rightValue) ^ negated;
        } 

        string toString() {
            string output = (negated ? "!" : "");
            return output + "(" + (left -> toString()) + " == " + (right -> toString()) + ")";
        }     
};

#endif

File: verification/src/main/java/org/msr/mnr/verification/expressions/EqExpr.java
package org.msr.mnr.verification.expressions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Arrays;

import org.msr.mnr.verification.dsfa.ConstraintTreeNode;
import org.msr.mnr.verification.dsfa.DSFA;
import org.msr.mnr.verification.utils.Packet;
import org.msr.mnr.verification.utils.ParseIntArray;


public class EqExpr extends BoolExpr {
    private static final long serialVersionUID = 3139893660332363038L;

    public EqExpr(Object left, Object right, HashSet<String> locationReferences,
            HashMap<String, String> variableComparisons) {
        super(left, right, locationReferences, variableComparisons);
    }

    @Override
    public Boolean evaluate(Packet p, ArrayList<int[]> constraints, ArrayList<String> locationList,
            ArrayList<String> variableList, Integer currentState, List<ConstraintTreeNode> rootConstraints) {

        // instantiate any packet fields. That way locations and variables are only
        // dealing with int[]s
        // System.out.println("ConditionA: " + this.toString());
        int[] leftValue = evalSingle(left, leftType, p, constraints, locationList, variableList);
        int[] rightValue = evalSingle(right, rightType, p, constraints, locationList, variableList);
        // System.out.println("Comparing: " + ParseIntArray.getString(leftValue) + " with: " + ParseIntArray.getString(rightValue));
        
        // Now parse variables.  Need to do this after the above so that we have the concrete int[] values
        if (leftType == ExprType.VARIABLE || leftType == ExprType.LOCATION_VAR) {
            int level;
            if (leftType == ExprType.VARIABLE) {
                level = locationList.size() + variableList.indexOf(left);
            } else {
                level = locationList.indexOf(left);
            }

            if (constraints.get(level) == null) {
                // Unconstrained, we may need to split, but only if we don't have an identical
                // set of constraints already
                assert(rightValue != null);
                DSFA.splitIfNew(constraints, 0, rightValue, currentState, rootConstraints);
                return negated; // normally false
            } else {
                // We have a constraint
                leftValue = constraints.get(level);
            }
        } else if (rightType == ExprType.VARIABLE || rightType == ExprType.LOCATION_VAR) {
            int level;
            if (rightType == ExprType.VARIABLE) {
                level = locationList.size() + variableList.indexOf(right);
            } else {
                level = locationList.indexOf(right);
            }

            if (constraints.get(level) == null) {
                // Unconstrained, we may need to split, but only if we don't have an identical
                // set of constraints already
                assert(leftValue != null);
                DSFA.splitIfNew(constraints, 0, leftValue, currentState, rootConstraints);
                return negated; // normally false
            } else {
                // We have a constraint
                rightValue = constraints.get(level);
            }
        } else if (leftValue == null || rightValue == null) {
            throw new RuntimeException("Unknown Eq Expression fields: " + left + " and " + right);
        }

        // equal ^ negated = false
        // equal ^ not negated = true
        // not equal ^ negated = true
        // not equal ^ not negated = false
        
        return (ParseIntArray.compare(leftValue, rightValue) == 0) ^ (negated);
    }

    @Override
    public Boolean evaluate(Packet p) {
        // instantiate any packet fields. That way locations and variables are only
        // dealing with int[]s
        // System.out.println("ConditionB: " + this.toString());
        int[] leftValue = evalSingle(left, leftType, p);
        int[] rightValue = evalSingle(right, rightType, p);
        // System.out.println("Comparing: " + Arrays.toString(leftValue) + " with: " + Arrays.toString(rightValue));
        // equal ^ negated = false
        // equal ^ not negated = true
        // not equal ^ negated = true
        // not equal ^ not negated = false
        return (ParseIntArray.compare(leftValue, rightValue) == 0) ^ (negated);
    }

    @Override
    public String toString() {
        if (rightType == ExprType.VALUE) {
            return (negated ? "!" : "") + "(" + left + " == " + Arrays.toString((int[]) right) + ")";
        } else {
            return (negated ? "!" : "") + "(" + left + " == " + right + ")";
        }
    }
}